bootable-jar/WFLY-14411_Bootable_Jar_Upgrade.adoc

```diff
@@ -0,0 +1,208 @@
+
+* Usage of Galleon feature-pack location (FPL) to reference the Galleon channel of a producer registered inside a universe) is still possible.
+
+* This mechanism strongly depends on Maven dependency resolution. 
+Any artifact to upgrade must be present in accessible Maven repositories (remote or local cache).
+
+* An artifact present in the  `overridden-artifacts` set must be present in the dependencies section. 
+It is advised that the dependency scope be `provided` to be not included in the application artifact (eg: war). 
+If an artifact is no present in the dependencies, then a failure occurs during build.
+
+* An artifact is referenced in the `overridden-artifacts` by GroupId, Artifactid and optionally Classifier. 
+Version being retrieved from the Maven dependencies. 
+
+* An  overridden artifact (artifact present in the `overridden-artifacts` list) must be unique. Any duplicate will make the packaging to fail.
+
+* Adding an overridden artifact that is not part of the provisioned server artifacts will lead to a failure during build.
+
+* Adding an overridden Galleon feature-pack artifact that is not a dependency of the WildFly server being 
+provisioned will lead to an error during packaging. 
+
+* The version of the galleon feature-packs referenced using GAV becomes optional in the maven plugin configuration if the  Galleon 
+feature-pack artifact has been added to the dependencies. In this case the version is being retrieved from the dependencies. If the version
+is missing and no Feature-pack artifact is retrieved in the dependencies, a failure occurs during packaging.
+
+* Third party galleon feature-packs benefit from this upgrade capability for JBoss Module artifacts they are bringing to the provisioned server.
+
+* The server jar components that can be upgraded are:
+** The JBoss module runtime jar (jboss-modules.jar file).
+** All jar artifacts referenced from JBoss Modules modules.
+
+* The set of artifacts that can be upgraded can be retrieved by setting the parameter `<dump-original-artifacts>true</dump-original-artifacts> or 
+the system property `bootable.jar.dump.original.artifacts` to true when building a bootable JAR. 
+The file `target/bootable-jar-build-artifacts/bootable-jar-server-original-artifacts.xml` is generated. 
+It contains XML elements for the Galleon feature-packs dependencies, JBoss Modules runtime and artifacts. 
+JBoss Modules modules artifacts are grouped by JBoss Modules name.
+The generated file contains only the artifacts that are provisioned by Galleon. 
+Each artifact version is the one that would get installed when building the Bootable JAR without upgrade.
+
+* An artifact upgraded to the same version as the one referenced in the Galleon feature-pack is not upgraded. In this case a warning is displayed during build.
+
+* It is possible to downgrade an artifact to an older version. In this case a warning is displayed during build.
```

Comment:
We do our best to detect downgrade. I suspect artifacts are versioned that way. For artifacts that don't comply we will warn and do the upgrade.

Reply:
@jfdenise That makes sense. What I was implying is that there might not be a reliable way how to recognize upgrade/downgrade - it can happen we have `1.2.3-m3` vs `1.2.3-SP4` (no one guarantees this won't happen). For sure we might be just `overriding` artifacts without explicitly stating whether it is an upgrade or downgrade. WDYT?

Reply:
I agree, that is not 100% accurate, but if the artifacts are properly versioned the way Maven specify it, we should we good. I think that warning the user that we detected it as downgrade is interesting (eg: for security reason).  Fo exotic versioning we could be wrong. We could introduce a configuration flag to disable the warning?

Reply:
Ah, yes, doing the best possible in this case makes sense. In my opinion the optimal solution would be if user would be able to list artifacts where downgrading will be "ignored". Though the configuration flag sounds good too!

Reply:
Thank-you, spec updated with the new flag.